planning_scene/src/planning_scene.cpp

```diff
@@ -1245,60 +1309,101 @@ bool planning_scene::PlanningScene::usePlanningSceneMsg(const moveit_msgs::Plann
     return setPlanningSceneMsg(scene_msg);
 }
 
-void planning_scene::PlanningScene::processOctomapMsg(const octomap_msgs::Octomap &map)
+void planning_scene::PlanningScene::removeAllCollisionObjects()
 {
-  // each octomap replaces any previous one
-  world_->removeObject(OCTOMAP_NS);
+  const std::vector<std::string> &object_ids = world_->getObjectIds();
+  for (std::size_t i = 0; i < object_ids.size(); ++i)
+    if (object_ids[i] != OCTOMAP_NS)
+      world_->removeObject(object_ids[i]);
+}
 
-  if (map.data.empty())
-    return;
+void planning_scene::PlanningScene::processOctomapMsg(const octomap_msgs::OctomapWithPose &msg)
+{
+  const Eigen::Affine3d &t = getTransforms().getTransform(msg.header.frame_id);
+  Eigen::Affine3d p;
+  tf::poseMsgToEigen(msg.origin, p);
+  p = t * p;
+  processOctomapMsg(msg.octomap, p);
+}
 
-  if (map.id != "OcTree")
+void planning_scene::PlanningScene::processOctomapMsg(const octomap_msgs::Octomap &msg)
+{
+  if (!msg.header.frame_id.empty())
   {
-    logError("Received octomap is of type '%s' but type 'OcTree' is expected.", map.id.c_str());
-    return;
+    const Eigen::Affine3d &t = getTransforms().getTransform(msg.header.frame_id);
+    processOctomapMsg(msg, t);
   }
+  else
+  {
+    processOctomapMsg(msg, Eigen::Affine3d::Identity());
+  }
+}
 
-  boost::shared_ptr<octomap::OcTree> om(static_cast<octomap::OcTree*>(octomap_msgs::msgToMap(map)));
-  if (!map.header.frame_id.empty())
+void planning_scene::PlanningScene::processOctomapMsg(const octomap_msgs::Octomap &msg, const Eigen::Affine3d &t)
+{
+  if (msg.id.empty())
   {
-    const Eigen::Affine3d &t = getTransforms().getTransform(map.header.frame_id);
+    world_->removeObject(OCTOMAP_NS);
+  }
+  else if (msg.id == OCTOMAP_DIFF_MSG_TYPE)
+  {
+    collision_detection::CollisionWorld::ObjectConstPtr map = world_->getObject(OCTOMAP_NS);
+    if (map && map->shapes_.size() == 1)
+    {
+      if (msg.data.empty())
+        return;
+      const shapes::OcTree *o = static_cast<const shapes::OcTree*>(map->shapes_[0].get());
+      boost::shared_ptr<const octomap::OcTree> octree = o->octree;
+      boost::shared_ptr<octomap::OcTree> nc_octree = boost::const_pointer_cast<octomap::OcTree>(octree);
```

Comment:
:-1:
removing const-ness in a context that is this complex invites disaster.
If you wish to be able to modify the octree, keep a non-const pointer in the planning scene.
This would also allow for some further simplifications...
However, this would break ABI and therefore would only be possible for kinetic-devel (I see dave jumping happily up and down when reading this :P).
As an alternative, you could simply copy the old octomap and drop it. I would still prefer the other solution.